### PR TITLE
✨clusterctl: improve move test framework

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR improves the move test framework addressing the requirements of the new "move by owner ref" approach and more specifically:

- all the objects now have a UID
- the UID is used in all the owner ref
- external objects are now Dummy types (instead of Unstructured)
- FakeProxy uses a new Fake schema including the Dummy types
- there are new methods for creating FakeCRDs
- owner Ref for template objects are aligned to changes introduced by https://github.com/kubernetes-sigs/cluster-api/pull/2007, https://github.com/kubernetes-sigs/cluster-api/pull/1947

**Which issue(s) this PR fixes**
Rif #1729

/assign @vincepri

### CODE REVIEW

File: cmd/clusterctl/pkg/internal/test/fake_objects.go [hunk 1]
```diff
@@ -775,15 +877,57 @@ func (f *FakeMachine) Objs(cluster *clusterv1.Cluster, generateCerts bool, machi
...
+		TypeMeta: metav1.TypeMeta{
+			Kind:       apiextensionslv1.SchemeGroupVersion.String(),
+			APIVersion: "CustomResourceDefinition",
 		},
-	})
-	externalRef.SetLabels(map[string]string{
-		clusterv1.ClusterLabelName: machine.Spec.ClusterName,
-	})
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("%s.%s", strings.ToLower(kind), group),
```
Comment (ncdc) [inline cmd/clusterctl/pkg/internal/test/fake_objects.go]: Technically the name needs to be pluralized if you want to match real CRD names. Do you want to do that?

File: Makefile [hunk 2]
```diff
@@ -189,7 +189,8 @@ generate-go-core: $(CONTROLLER_GEN) $(CONVERSION_GEN)
 	$(CONTROLLER_GEN) \
 		object:headerFile=./hack/boilerplate/boilerplate.generatego.txt \
 		paths=./api/... \
-		paths=./cmd/clusterctl/api/...
+		paths=./cmd/clusterctl/api/... \
+		paths=./cmd/clusterctl/pkg/internal/test/providers/...
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply Makefile]: done. this implies a full scan of the clusterctl codebase, but I guess it is fine.

File: cmd/clusterctl/pkg/internal/test/providers/bootstrap/dummy_types.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+*/
+package bootstrap
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type DummyBootstrapConfigStatus struct {
+	// +optional
+	DataSecretName string `json:"dataSecretName,omitempty"` //TODO: capire perche con *string fallisce...
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/internal/test/providers/bootstrap/dummy_types.go]: The field is `*string` in the KubeadmConfig type, so I would like to do the same here.
But then it was giving me some troubles, so I removed the pointer temporarily but then forgot to fix it 😊 .
Now I have restored the pointer to respect the KubeadmConfig example/the cluster API spec

File: cmd/clusterctl/pkg/internal/test/providers/bootstrap/dummy_types.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []DummyBootstrapConfig `json:"items"`
+}
+
+// +kubebuilder:object:root=true
+
+type DummyBootstrapConfigTemplate struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/internal/test/providers/bootstrap/dummy_types.go]: Using the real field comes handy, it matches what we have in KubeadmBootstrap and KubeadmControlPlane providers, and it makes less painful all the conversions from Unstructured/Typed that are spread into the code.

FYI, atm in the tests I'm using `Name`, `Namespace`, `Labels` and `OwnerReference` plus some utils like `SetOwnerReferences`. Additionally, those types are processed by `controller-gen` to generate the deepcopy stuff.

File: cmd/clusterctl/pkg/internal/test/providers/controlplane/dummy_types.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+type DummyControlPlaneSpec struct {
+	InfrastructureTemplate corev1.ObjectReference `json:"infrastructureTemplate"`
+}
+
+// +kubebuilder:object:root=true
+
+type DummyControlPlane struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              DummyControlPlaneSpec `json:"spec,omitempty"`
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/internal/test/providers/controlplane/dummy_types.go]: [1] Currently, the types have the minimum set of fields that are required to test move.
If we want to expand this set and implements all the fields that are defined in the cluster API contract I'm ok with this, but if possible, I would address that in a separated PR

File: cmd/clusterctl/pkg/internal/test/providers/infrastructure/dummy_types.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +kubebuilder:object:root=true
+
+type DummyInfrastructureCluster struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/internal/test/providers/infrastructure/dummy_types.go]: [1]

File: cmd/clusterctl/pkg/internal/test/providers/infrastructure/dummy_types.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []DummyInfrastructureCluster `json:"items"`
+}
+
+// +kubebuilder:object:root=true
+
+type DummyInfrastructureMachine struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/internal/test/providers/infrastructure/dummy_types.go]: [1]

File: cmd/clusterctl/pkg/internal/test/providers/infrastructure/dummy_types.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []DummyInfrastructureMachine `json:"items"`
+}
+
+// +kubebuilder:object:root=true
+
+type DummyInfrastructureMachineTemplate struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/internal/test/providers/infrastructure/dummy_types.go]: [1]

File: cmd/clusterctl/pkg/internal/test/providers/controlplane/groupversion_info.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// +kubebuilder:object:generate=true
+// +groupName=controlplane.cluster.x-k8s.io
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/internal/test/providers/controlplane/groupversion_info.go]: [2] No, I don't think this is required. On the contrary, I removed `.clusterctl` from the infra group so we are consistent with the KubeadmBootstrap, KubeadmControlPlane and the Docker providers.

File: cmd/clusterctl/pkg/internal/test/providers/bootstrap/groupversion_info.go [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// +kubebuilder:object:generate=true
+// +groupName=bootstrap.cluster.x-k8s.io
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/internal/test/providers/bootstrap/groupversion_info.go]: [2]

File: cmd/clusterctl/pkg/internal/test/fake_objects.go [hunk 11]
```diff
@@ -775,15 +877,57 @@ func (f *FakeMachine) Objs(cluster *clusterv1.Cluster, generateCerts bool, machi
...
-func reconcileMachineExternalRef(externalRef *unstructured.Unstructured, machine *clusterv1.Machine) {
-	externalRef.SetOwnerReferences([]metav1.OwnerReference{
-		{
-			APIVersion: machine.APIVersion,
-			Kind:       machine.Kind,
-			Name:       machine.Name,
+// setUID assigns a UID to the object, so test objects are uniquely identified.
+// NB. In order to make debugging easier we are using a human readable, deterministic string (instead of a random UID).
+func setUID(obj runtime.Object) {
+	accessor, _ := meta.Accessor(obj)
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/internal/test/fake_objects.go]: done

File: cmd/clusterctl/pkg/internal/test/fake_objects.go (hunk 1 above)
Comment (fabriziopandini) [reply cmd/clusterctl/pkg/internal/test/fake_objects.go]: For the sake of test what really matters is to generate a unique name, so I think we can leave as it is

**Comment by fabriziopandini** [comment]
@ncdc @vincepri many thanks for the feedback!
I addressed some of them and provided answer to the questions

File: cmd/clusterctl/pkg/internal/test/providers/bootstrap/dummy_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ncdc) [reply cmd/clusterctl/pkg/internal/test/providers/bootstrap/dummy_types.go]: Sorry, my question wasn't clear. Do you want to add any spec/status fields?

File: cmd/clusterctl/pkg/internal/test/providers/controlplane/dummy_types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ncdc) [reply cmd/clusterctl/pkg/internal/test/providers/controlplane/dummy_types.go]: That's fine, I'm all good for keeping things simple.

File: cmd/clusterctl/pkg/internal/test/providers/bootstrap/dummy_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ncdc) [reply cmd/clusterctl/pkg/internal/test/providers/bootstrap/dummy_types.go]: Saw you're answer below. All good!

**Comment by ncdc** [comment]
LGTM! Will need squashing once @vincepri is done reviewing.
/approve

**Comment by vincepri** [comment]
LGTM pending squash

**Comment by fabriziopandini** [comment]
@vincepri squashed
Thanks!

**Pull request merged**